packages/icon/package.py

```diff
@@ -1,294 +1,710 @@
+    variant('jsbach', default=True, description='Enable the land component')
+    variant('coupling', default=True, description='Enable the coupling')
+    variant('ecrad',
             default=False,
-            description='Build with claw directories enabled')
+            description='Enable usage of the ECMWF radiation scheme')
     variant('rte-rrtmgp',
-            default=True,
-            description='Build with rte-rrtmgp enabled')
-    variant('mpi-checks',
             default=False,
-            description='Build with mpi-check enabled')
-    variant('openmp', default=True, description='Build with openmp enabled')
-    variant('serialize_mode',
-            default='none',
-            description='Build with serialization, with serialze_mode enabled',
-            values=('none', 'create', 'read', 'perturb'))
-    variant('eccodes', default=False, description='Build with grib2 enabled')
-    variant('test_name',
-            default='none',
-            description='Launch test: test_name after installation')
-    variant('skip-config', default=False, description='Skip configure phase')
-    variant('config_dir',
-            default='.',
-            description='Enable out-of-source build by describing config_dir')
-    variant('ham',
+            description='Enable usage of the RTE+RRTMGP toolbox '
+            'for radiation calculations')
+    variant(
+        'rttov',
+        default=False,
+        description='Enable usage of the radiative transfer model for TOVS')
+    variant('dace',
             default=False,
-            description='Build with hammoz and atm_phy_echam enabled.')
-    variant('art', default=False, description='Build with art enabled')
-    variant('ocean', default=True, description='Build with ocean enabled')
-    variant('dace', default=False, description='Build with DACE enabled')
-    variant('rttov', default=False, description='Build with RTTOV enabled')
-    variant('mixed-precision',
+            description='Enable the DACE modules for data assimilation')
+    variant('emvorado',
+            default=False,
+            description='Enable the radar forward operator EMVORADO')
+    variant('art',
             default=False,
-            description='Build the ICON dycore in mixed precision')
-    variant('silent-rules',
+            description='Enable the aerosols and reactive trace component ART')
+
+    # Infrastructural Features:
+    variant('mpi',
             default=True,
-            description='Build with Make silent rules ON')
+            description='Enable MPI (parallelization) support')
+    variant('openmp', default=False, description='Enable OpenMP support')
+
+    # https://en.wikipedia.org/wiki/CUDA#GPUs_supported
+    gpu_values = ('10', '11', '12', '13', '20', '21', '30', '32', '35', '37',
+                  '50', '52', '53', '60', '61', '62', '70', '72', '75', '80',
+                  '86')
+    variant('gpu',
```

Comment:
This is a leaky level of abstraction, because sysconfig isn't able to abstract the machine now. The machine specific detail, which cuda_arch/sm is needed on that machine, is required outside of sysconfig to build a cuda accelerated variant.
Spack packages commonly inherit from CudaPackage to enable cuda compatibility. I started implementing this some time ago here: https://github.com/C2SM/spack-c2sm/blob/new_icon/packages/icon/package.py but it's not tested thoroughly.

Reply:
CudaPackage comes with the additional benefit, that we don't need to take care of the list of gpu_values and the icon package is forward compatible with spack v0.19's feature `++cuda` and `~~cuda`, which means, the entire subtree should have `+cuda` or `~cuda`.

Reply:
Could we move this discussion to another PR?

I really like what you propose but I don't want to blow up this branch even more.